internal/pkg/otel/manager/extension.go

```diff
@@ -0,0 +1,185 @@
+	ctx, cancel := context.WithCancel(ctx)
+	aggregator := status.NewAggregator(status.PriorityRecoverable)
+	as := &AgentStatusExtension{
+		mgr:        mgr,
+		telemetry:  set.TelemetrySettings,
+		aggregator: aggregator,
+		eventCh:    make(chan *evtPair),
+		readyCh:    make(chan struct{}),
+		kickCh:     make(chan struct{}, 1),
+		ctx:        ctx,
+		canceller:  cancel,
+	}
+
+	// start processing early as ComponentStatusChanged will be called before Start is called
+	go as.eventLoop(ctx)
+
+	return as
+}
+
+// NewAgentStatusFactory provides a factory for creating the AgentStatusExtension.
+func NewAgentStatusFactory(mgr *OTelManager) extension.Factory {
+	return extension.NewFactory(
+		AgentStatusExtensionType,
+		func() component.Config {
+			return nil
+		},
+		func(ctx context.Context, set extension.Settings, cfg component.Config) (extension.Extension, error) {
+			return NewAgentStatusExtension(ctx, set, mgr), nil
+		},
+		component.StabilityLevelDevelopment,
+	)
+}
+
+// Start implements the component.Component interface.
+func (as *AgentStatusExtension) Start(ctx context.Context, host component.Host) error {
+	as.telemetry.Logger.Debug("Starting agent status extension")
+	as.host = host
+	return nil
+}
+
+// Shutdown implements the component.Component interface.
+func (as *AgentStatusExtension) Shutdown(ctx context.Context) error {
+	// preemptively send the stopped event, so it can be exported before shutdown
+	componentstatus.ReportStatus(as.host, componentstatus.NewEvent(componentstatus.StatusStopped))
+	as.canceller()
+	return nil
+}
+
+// Ready implements the extension.PipelineWatcher interface.
+func (as *AgentStatusExtension) Ready() error {
+	close(as.readyCh)
+	return nil
+}
+
+// NotReady implements the extension.PipelineWatcher interface.
+func (as *AgentStatusExtension) NotReady() error {
+	return nil
+}
+
+// ComponentStatusChanged implements the extension.StatusWatcher interface.
+func (as *AgentStatusExtension) ComponentStatusChanged(
+	source *componentstatus.InstanceID,
+	event *componentstatus.Event,
+) {
+	// this extension is always force loaded and not by the user, so status
+	// information should be hidden as they didn't directly enable it
+	if source.ComponentID().String() == AgentStatusExtensionType.String() {
+		return
+	}
+	// possible that even after Shutdown is called that this function is still
+	// called by the coordinator
+	defer func() {
+		if r := recover(); r != nil {
+			as.telemetry.Logger.Info(
+				"discarding event received after shutdown",
+				zap.Any("source", source),
+				zap.Any("event", event),
+			)
+		}
+	}()
+	select {
+	case as.eventCh <- &evtPair{source: source, event: event}:
+	case <-as.ctx.Done():
+	}
+}
+
+func (as *AgentStatusExtension) eventLoop(ctx context.Context) {
+	// prevent aggregate statuses from flapping between StatusStarting and StatusOK
+	// as components are started individually by the service.
```

Comment:
Could we just not do this filtering? Was it just annoying or causing an actual problem?

Reply:
It causes the reporting to not really be in sync and looks very strange. This is also following the same pattern of the healthcheckv2extension.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/healthcheckv2extension/extension.go#L168

Reply:
Ah, can you link to that code directly?

Reply:
Done.